source/extensions/filters/http/common/aws/utility.h

```diff
@@ -39,10 +39,26 @@ class Utility {
    */
   static std::string
   joinCanonicalHeaderNames(const std::map<std::string, std::string>& canonical_headers);
+
+  /**
+   * Fetch AWS instance or task metadata.
+   *
+   * @param host host or ip address of the metadata endpoint.
+   * @param path path of the metadata document.
+   * @auth_token authentication token to pass in the request, empty string indicates no auth.
+   * @return Metadata document or nullopt in case if unable to fetch it.
+   *
+   * @note In case of an error, function will log ENVOY_LOG_MISC(debug) message.
+   *
+   * @note This is not main loop safe method as it is blocking. It is intended to be used from the
```

Comment:
Why use curl here at all and have it be blocking? Why can't you use the built in Envoy HTTP async client facilities?

Reply:
We initially implemented this using the Envoy Http client in #5546. Although I do not know what was the original motivation, there was a suggestion on that PR to use libcurl.

@htuch  and @lavignes can comment more on this.

Reply:
Yeah. Don't have much more to add. When I originally implemented this, I ran into the issue that I ultimately needed to block signing the xDS request on the http request to get credentials. Note that the blocking is done on the gRPC client thread in my implementation. Not the main thread. It was after asking about methods for making blocking http requests it was suggested I make the blocking http client in #5546. But I don't recall in what issue it was suggested.

Then later in #5546 just using curl was proposed since it was being merged in as a dependency anyway. I felt more comfortable doing that since I wasn't very confident in my http client implementation. It was non-trivial for me to implement and test especially since I was very new to the codebase.

Reply:
OK that's fine. Ultimately I don't think curl should be used here for a variety of reasons (blocking, more third party code to deal with, not the normal Envoy logs/metrics/traces, etc.) but we can merge and iterate.

Reply:
I think I pointed out that libcurl was now available, but I can't remember if this was something we decided as the preferred approach. The main concern would be that we don't block on a worker thread either in the Envoy-based HTTP client or libcurl.